src/features/images/tasks/delete-registry-images.ts

```diff
@@ -0,0 +1,176 @@
+			type: 'array',
+			items: {
+				type: 'array',
+				items: {
+					type: 'string',
+				},
+				maxItems: 2,
+				minItems: 2,
+			},
+		},
+	},
+	required: ['images'],
+};
+
+export type DeleteRegistryImagesTaskParams = {
+	images: Array<[repo: string, hash: string]>;
+};
+
+const handlerName = 'delete_registry_images';
+const logHeader = handlerName.replace(/_/g, '-');
+tasks.addTaskHandler(
+	handlerName,
+	async (options) => {
+		try {
+			const images =
+				(options.params as DeleteRegistryImagesTaskParams).images ?? [];
+			if (images.length === 0) {
+				return {
+					status: 'succeeded',
+				};
+			}
+
+			// Chunk by batch size in case we need to tune after tasks have been created
+			for (const chunk of _.chunk(
+				images,
+				ASYNC_TASK_DELETE_REGISTRY_IMAGES_BATCH_SIZE,
+			)) {
+				// Avoid deleting any blobs that are still referenced by other images
+				// This shouldn't normally be necessary as is_stored_at__image_location
+				// should be enforced as unique at the database level, but just in case
+				const stillReferenced = await api.resin.get({
+					resource: 'image',
+					passthrough: { req: permissions.rootRead },
+					options: {
+						$select: ['id', 'is_stored_at__image_location', 'content_hash'],
+						$filter:
+							chunk.length === 1
+								? {
+										is_stored_at__image_location: chunk[0][0],
+										content_hash: chunk[0][1],
+									}
+								: {
+										$or: chunk.map(([repo, hash]) => ({
+											is_stored_at__image_location: repo,
+											content_hash: hash,
+										})),
+									},
+					},
+				});
+				const safeToDelete = chunk.filter(
+					([repo, hash]) =>
+						!stillReferenced.some(
+							(image) =>
+								image.is_stored_at__image_location === repo &&
+								image.content_hash === hash,
+						),
+				);
+				for (const [repo, hash] of safeToDelete) {
+					if (repo === '' || hash === '') {
+						console.warn(
+							`[${logHeader}] Skipping deletion of image with empty repo or hash: ${repo}/${hash}`,
+						);
+						continue;
+					}
+					await markForDeletion(repo, hash);
+				}
+			}
+
+			return {
+				status: 'succeeded',
+			};
+		} catch (e) {
+			console.error(`[${logHeader}] Error marking images for deletion: ${e}`);
+			return {
+				error: `${e}`,
+				status: 'failed',
+			};
+		}
+	},
+	schema,
+);
+
+// Make an API call to the registry service to mark images for deletion on next garbage collection
+async function markForDeletion(repo: string, hash: string) {
+	// Generate an admin-level token with delete permission
+	const token = generateToken('admin', REGISTRY2_HOST, [
+		{
+			name: repo,
+			type: 'repository',
+			actions: ['delete'],
+		},
+	]);
+
+	// Need to make requests one image at a time, no batch endpoint available
+	for (let retries = 0; retries < RATE_LIMIT_RETRIES; retries++) {
+		const [{ statusCode, statusMessage, headers }] = await requestAsync({
+			url: `https://${REGISTRY2_HOST}/v2/${repo}/manifests/${hash}`,
```

Comment:
Are all repo/hashes we have here guaranteed to belong to our own registry or could these be images on custom customer registry?